paperqa/llms.py

```diff
@@ -764,6 +802,378 @@ async def select_tool(
+        }
+
+        for i, d in enumerate(data):
+            batch_template["custom_id"] = str(i)
+            batch_template["body"]["model"] = self.config.get("model")
+            batch_template["body"]["messages"] = d
+            batch_template["body"]["max_tokens"] = self.config.get("max_tokens")
+            serialized_data = json.dumps(batch_template) + "\n"
+            tmp_file.write(serialized_data.encode())
+            # tmp_file.write(json.dumps(batch_template) + "\n")
+
+    async def acomplete(self):
+        raise NotImplementedError("Only chat models are supported by openAI batch API.")
+
+    async def acomplete_iter(self):
+        raise NotImplementedError(
+            "Async generator not supported for batch calls and nly chat models are supported by openAI batch API."
+        )
+
+    async def _run_chat(
+        self,
+        prompt: str,
+        data: list[dict],
+        callbacks: list[Callable] | None = None,
+        name: str | None = None,
+        system_prompt: str | None = default_system_prompt,
+    ) -> list[LLMResult]:
+        if callbacks:
+            sync_callbacks = [f for f in callbacks if not is_coroutine_callable(f)]
+            async_callbacks = [f for f in callbacks if is_coroutine_callable(f)]
+
+        human_message_prompt = {"role": "user", "content": prompt}
+
+        batch = []
+        for d in data:
+            messages = [
+                {"role": m["role"], "content": m["content"].format(**d)}
+                for m in (
+                    [{"role": "system", "content": system_prompt}, human_message_prompt]
+                    if system_prompt
+                    else [human_message_prompt]
+                )
+            ]
+            batch.append(messages)
+
+        start_clock = asyncio.get_running_loop().time()
+        chunks = await self.achat(batch)
+        batch_time = asyncio.get_running_loop().time() - start_clock
+
+        if callbacks:
+            for chunk in chunks:
+                await do_callbacks(
+                    async_callbacks, sync_callbacks, chunk.text or "", name
+                )
+
+        return [
+            LLMResult(
+                model=self.name,
+                name=name,
+                prompt=messages,
+                prompt_count=chunk.prompt_tokens,
+                text=chunk.text,
+                completion_count=chunk.completion_tokens,
+                seconds_to_first_token=batch_time,
+                seconds_to_last_token=batch_time,
+            )
+            for messages, chunk in zip(batch, chunks, strict=True)
+        ]
+
+    async def achat(self, messages: list[list[dict]]) -> list[Chunk]:
+        try:
+            import openai
+        except ImportError as exc:
+            raise ImportError(
+                "Please install paper-qa[batch] to use OpenAIBatchLLMModel."
+            ) from exc
+
+        client = openai.AsyncOpenAI()
+
+        with io.BytesIO() as tmp_file:
```

Comment:
This isn't a `tmp_file` (it's not a file in the `/tmp` dir), let's call it `mem_buffer` or `mem_file`